### DISCUSSION TITLE
[ZK Stack] -How to use the basetoken feature to deploy and use customer native token?

### DISCUSSION THREAD

**Discussion by SuccinctPaul** [zkSync-Community-Hub/zksync-developers#622] [ZK Stack] [answered]
### Module Affected

- [ ] System contracts (era-system-contracts)
- [ ] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [x] Server (zksync-era)
- [ ] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [x] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [ ] Other (please specify below)

### Rust Version

1.75.0

### Operating System

- [X] Ubuntu
- [ ] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

How to use basetoken feature to set customer native l2's native token?
And how to bridge the token from l1?

### Expected Behavior

Step by step tutorial to deploy and use customer native token?

### Current Behavior

No idea to do this.

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by alhenaai** [comment]
To deploy and use a custom native token as the base token on a zkSync L2 chain and bridge it from L1, you can follow these steps based on the zkSync documentation and system contracts:

### Step 1: Deploy Your ZK Chain
First, you need to deploy your own ZK Chain. This involves specifying the governance, STM (State Transition Machine), and the baseToken you want to use. Here’s how you can do it:

1. **Inform zkSync**: Typically, you would inform zkSync (possibly through a form or direct communication) that you want to deploy a ZK Chain with specific configurations including the baseToken.
2. **Chain Deployment**: zkSync's server will generate a unique chainId not reserved by any other major chain. The admin will then call `BridgeHub.createNewChain`, which in turn calls `STM.createNewChain`. This deploys the rollup instance and initializes it with the first transaction — the system upgrade transaction needed to set the chainId on L2.
3. **Chain Readiness**: After these steps, your ZK Chain is ready for use.

### Step 2: Using the BaseToken
The baseToken you choose will be used as the native token on your L2 chain. Here’s how to manage and use it:

1. **SharedBridge Contract**: The SharedBridge contract on L1 will store all the funds, including your baseToken. This contract is crucial as it maintains the balance of all baseTokens, including all of the ETH supply if ETH is chosen as a baseToken.
2. **L2BaseToken Contract**: On L2, the L2BaseToken contract will manage the baseToken. It handles the balances of the baseToken for users but does not provide an ERC20 interface. The only method for transferring the baseToken is through `transferFromTo`, which permits only some system contracts to transfer on behalf of users.

### Step 3: Bridging the Token from L1 to L2
To bridge your custom native token from L1 to L2, follow these steps:

1. **Deposit via Bridge**: Use the `zksync-cli` tool to facilitate bridge operations between L1 and L2. Here’s a basic command to deposit your token from L1 to L2:
   ```bash
   zksync-cli bridge deposit --to [recipient L2 address] --amount [amount] --token [L1 token address] --chain [chain] --l1-rpc [L1 RPC URL]
   ```
2. **Manage Tokens on L2**: Once bridged, the L2SharedBridge on L2 will mint and manage the corresponding tokens. Users can interact with these tokens using standard L2 operations.

### Step 4: Additional Configurations and Testing
- **Testing**: Ensure to test the entire flow on a testnet before moving to production.
- **Customization**: Depending on your requirements, further customization might be needed, especially if you plan to support additional features or different types of tokens.

### Documentation and Resources
- For more detailed information and updates, refer to the [zkSync Documentation](https://docs.zksync.io/).
- For specific contract interactions and examples, check the [GitHub repository](https://github.com/matter-labs/era-contracts) of zkSync contracts.

This guide provides a high-level overview, and specific details may vary based on the exact requirements and updates to the zkSync platform. Always ensure to stay updated with the latest documentation and community discussions.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by sarahschwartz** [comment, marked as answer]
Hi @SuccinctPaul, the general flow for using a custom base token assuming you are running everything locally and have already gone through this guide: https://docs.zksync.io/zk-stack/running-a-zk-chain/locally

- start the Docker containers so the local reth node is running with `zk_inception containers`
- deploy an ERC20 token token to the L1
- send ERC20 tokens to the governor address on L1 (can be found in `path/to/ecosystem_folder/configs/wallets.yaml`)
- create your new chain that uses the ERC20 token & set it as the default
- initialize the ecosystem with the new default chain
- bridge tokens from L1 to L2 using `zksync-cli bridge`: https://docs.zksync.io/build/zksync-cli/bridging

A step-by-step tutorial for this will be coming soon!
